### ISSUE TITLE
Makeflow Should Perform Local Resource Management

### ISSUE THREAD

**Issue by zyli5313** [cooperative-computing-lab/cctools#1695]
Hi ,

I have a makeflow pipeline and in the middle of a pipeline I have, for example, p1 and p2 running in parallel. However, p1 and p2 both use large memory that are too big to fit the in memory if they run in parallel. But it's ok I run them one by one. 

I specified MEMORY=7168 at the start of the makeflow script. But it seems this only means makeflow will allocate 7.1Gb but it doesn't say to streamline these 2 processes. My question is is there a way to streamline processes if they're expensive in memory?

Thanks!

**Comment by btovar** [comment]
Which batch system are you using?

**Comment by zyli5313** [comment]
I was using local machine. No batch system.

**Comment by zyli5313** [comment]
I tried to add MEMORY=6000 each for two parallel jobs on a 8GB server. It still failed with memory error.

Is there a way to solve this? I can setup torque or slurm if that support this function.

**Comment by btovar** [comment]
Yes, it seems that makeflow leaves the resource control to the batch systems, and we have not extended such control to local execution. There is a way in which you can still run locally, using work queue as a batch system:

makeflow -Twq -p YYYY yourmakeflowfile

where YYYY is some free port that the machine has. A common default for work queue is 9123.

In another terminal, on the same machine, type:
work_queue_worker  --cores  ??? --disk ??? --memory 7000 localhost 9123

Replace the ??? with appropriate values. With that, the worker will execute as many tasks that fit in 7000 (minus some memory for bookkeeping, etc.).

Let us know if that works for you! If it does, I can give you another recipe, where you would not have to keep track of the port numbers used, which is useful for when you have many workers running (e.g., your workers may be running on slurm or torque).

**Comment by dthain** [comment]
That's a good workaround, but we should really have Makeflow do some resource management when running locally.  Jake won't be the only person to run into this problem..

**Comment by dthain** [comment]
Duplicate of issue #254

**Comment by dthain** [comment]
Working on this in PR #1697

**Comment by zyli5313** [comment]
Thanks for the quick response. I tried "makeflow -Twq -p 9123 example.mf" but my .mf script contains absolute path which is not supported by work queue. I will try installing slurm. Thanks!

**Comment by btovar** [comment]
Alternatively, if you don't mind testing our new additions in: 

https://github.com/cooperative-computing-lab/cctools/pull/1698 a try.

You can specify  makeflow --local-memory 7000 yourmakeflow

**Comment by dthain** [comment]
It's not quite merged yet, I'll let you know when it's ready to try...

**Comment by dthain** [comment]
@zyli5313 please check out the latest makeflow source from github, and try your original example again, using local execution.  Let us know if that fixes your problem.

**Comment by zyli5313** [comment]
Thanks! Fast turnaround! @btovar @dthain . 

One more things to clear. If I specify .mf files like this

ORES=4
MEMORY=7168
DISK=8192
outA: inA
python taskA.py 

outB: inB
python taskB.py

Is only taskA got limited by 4 cores, 7g memory and at the same time taskB is free of this constraint? Or both tasks are governed by 4 cores, 7g memory limit?

**Comment by zyli5313** [comment]
I tried just now (You can specify makeflow --local-memory 7000 yourmakeflow). And I specified 
CORES=4
MEMORY=7168
DISK=8192
at the front of the example.mf file. All my tasks seems to run one by one, which is in contrast with that before. It used to run parallel. I was aiming to run everything in parallel except to run those small number of memory constraint task one by one. Did I miss anything? Thanks!

**Comment by btovar** [comment]
Without anything else, the resources are applied to all the task. I think you want something like this:

.MAKEFLOW CATEGORY small-tasks
.MAKEFLOW CORES 1
a: b
      somecmd

...(some other small tasks.)...

.MAKEFLOW CATEGORY big-tasks
.MAKEFLOW CORES 4
.MAKEFLOW MEMORY 7168
.MAKEFLOW DISK 8192
c: d
    someothercmd

...(some other big tasks)...


.MAKEFLOW CORES X  is the same as CORES=X, but we like .MAKEFLOW better.
Note that I assign some resources to the small tasks, otherwise they all may run together with the big tasks. That may or may not be what you want.

Without assigning categories, all your tasks and resources were assigned to the default category. 

Ben

**Comment by zeronot** [comment]
@zyli5313  how do you know the resource leads to the  memory errors ? it seems that makeflwo do not provide any errors information.

**Comment by dthain** [comment]
Closing this issue b/c the feature is now implemented and I believe the problem is fixed.  But let us know if you have any trouble.
